Max sample rate test (#181)
SW709E19/Report: report/ChapterTest/StressTest.tex

> \section{Max Sample Rate Test}

Reviewer: ```suggestion
\section{Max Sample Rate Test}\label{sec:MaxSampleRateTest}
```

\subsubsection{MPU-6050}
The MPU-6050 is rated to run on different settings, that measure different ranges.
The output rate is programmable for both gyroscope and accelerometer.
The gyroscope can run at anything from 4 Hz to 8 KHz, and the accelerometer can run at anything from 4 Hz to 1 KHz. \cite{Mpu-6050-specifications} 
The gyroscope is rated for measuring angular velocity measured in degrees per second in different ranges, with varying accuracy. The gyroscope supports the follow ranges \cite{Mpu-6050-specifications} 
\begin{itemize}
 \item $\pm 250 ~^{\circ} / s$
 \item $\pm 500 ~^{\circ} / s$
 \item $\pm 1000 ~^{\circ} / s$
 \item $\pm 2000 ~^{\circ} / s$
\end{itemize}
\textit{Note: Selecting a higher range, will give a larger range, but less accuracy \cite{Mpu-6050-specifications}.}

The accelerometer is rated for measuring gravitational force in different ranges, it supports the following \cite{Mpu-6050-specifications} 
\begin{itemize}
 \item $\pm 2 ~g$
 \item $\pm 4 ~g$
 \item $\pm 8 ~g$
 \item $\pm 16 ~g$
\end{itemize}
\textit{Note: Selecting a higher range, will give a larger range, but less accuracy \cite{Mpu-6050-specifications}.}

> The MPU-6050 is able to run in two different output modes to provide data over $I^2C$ \cite{Mpu-6050-specifications}. 

Reviewer 2: Jeg tror at vi skriver I2C uden kursiv andre steder. Dvs I$^2$C. Og vi skriver det også uden 2-tallet opløftet i existing solutions. Er det linteren vi skal have fat i? ;)

> \begin{itemize}

Reviewer 2: Er det nødvendigt at have disse værdier hele 3 gange? Er denne første gang ikke nok?

>  \item Standard mode - 100 KHz - 100 KBits/s

Reviewer 2: Hvis du er enig i denne rettelse må du gerne lige rette alle de andre steder også
```suggestion
 \item Standard mode - 100 KHz - 100 kbits/s
```

 \hline
\textbf{Sample Rate} &\textbf{Measurements} &\textbf{Expected} &\textbf{Relative error}\\ \hline
 100Hz &500 &500 & 0,00\% \\ \hline
 200Hz &995 &1000 &-0,50\%\\ \hline
 300Hz &1485 &1500 &-1,00\%\\ \hline
 400Hz &1984 &2000 &-0,80\%\\ \hline
 500Hz &2438 &2500 &-2,50\%\\ \hline
 600Hz &2899 &3000 &-3,40\%\\ \hline
 700Hz &3404 &3500 &-2,70\%\\ \hline
 800Hz &3800 &4000 &-5,00\%\\ \hline
 900Hz &4300 &4500 &-4,40\%\\ \hline
1000Hz &4813 &5000 &-3,70\%\\ \hline
1100Hz &5170 &5500 &-6,00\%\\ \hline
1200Hz &5619 &6000 &-6,40\%\\ \hline
1300Hz &6045 &6500 &-7,00\%\\ \hline
1400Hz &6649 &7000 &-5,00\%\\ \hline
1500Hz &6891 &7500 &-8,10\%\\ \hline
1600Hz &7190 &8000 &-10,10\%\\ \hline
1700Hz &7558 &8500 &-11,10\%\\ \hline
1800Hz &7583 &9000 &-15,70\%\\ \hline
1900Hz &7674 &9500 &-19,20\%\\ \hline
2000Hz &7382 &10000 &-26,20\%\\ \hline
\end{tabular}
\caption{Raw data from $I^2C$ Fast mode}
\label{tab:fast-mode-table}
\end{table}

\subsection{Conclusion}
As can be seen on \autoref{fig:sample-rate-stress-test}, the Standard mode starts to have trouble to handle the amount of data at a sample rate of 500 Hz, while Fast mode starts to have trouble keeping up at around 1500 Hz.

> This means that Fast mode can handle more data than Standard mode, as it is expected since Fast mode has 4 times the throughput of 400 KBit/s compared to Standard modes 100 KBit/s.

Reviewer 2: Hvis du er uenig skal du bare lige tilføje lave "modes => mode's"
```suggestion
This means that Fast mode can handle more data than Standard mode, as it is expected since Fast mode has 4 times the throughput.
```

Since the Fast mode can't handle 4 times the sample rate of Standard mode, then there must be a bottleneck other than the limitations of the $I^2C$ protocol.
In this case, since the MPU-6050 is rated to output at 8 kHz, and the highest sample rate tested is just 2 kHz, the bottleneck is not with the MPU-6050.
In \autoref{fig:relative-error-percentage} the relative error percentage can be seen at the different sample rates.

> It is visible that when the limitation of the $I^2C$ protocols Standard mode is reached, the error percentage quickly increases. 

Reviewer 2: Jeg kan ikke få det til at give mening med de grænseværdier vi viste for I2C. Men jeg har nok bare misforstået noget?

But it is not the same steep slope when the fast mode error percentage starts to increase.
This could indicate that it is not the same bottleneck for Fast mode and Standard mode.
The test does not directly indicate where the problem is, but it could be speculated that since the ESP-32 has to read the data and store it, the ESP-32 is the bottleneck.
A possible solution to this is to use a more powerful microcontroller, but this would possibly increase the cost of the product.

> This means that the max sample rate the system is capable to handle is around 1500 Hz, which will be sufficient for this project.

Reviewer 2: Men vil det ikke også have en effekt på batterilevetid hvis vi vælger fast mode over standard mode?

Author: Jo det vil det nok men vi har ikke testet det, så det er svært at sige

%Stress test
% Theoretical maximum
% What is the MPU rated for
% What should the ESP be able to handle
% Wi-Fi speed
% Interrupts
% processing power
% Storage
% Maximum transfer speed of ESP over Wi-Fi to the Raspberry pi
% Raspberry Pi rating
% ESP rating
% (surprise ESP is slow)
% Actual maximum
% How fast can the ESP read from the MPU
% What is the limiting factor
% How fast can the ESP send data to the raspberry pi
% What is the max throughput
% Results from Mortens stress test
% What are the limiting factors
% Signal strength
% Distance from ESP to Gateway
% Obstacles (Walls, Machines, other Wi-Fi noise)
% ESP Cache is full and it has to offload data
% Limited memory and storage
% Conclusion
% ESP is the bottleneck
% Wi-Fi is relatively slow
% There is very little storage to store values in
% Limited processing power

This section describes a test of the ESP32 and MPU-6050 devices to find out how much they can handle in order to determine the system's upper sample rate limit. 

\subsection{Theoretical Maximum}\label{subsec:theoretical-maximum}
In this section, the theoretical maximum of the different components of the system will be researched.
\subsubsection{MPU-6050}

> The MPU-6050 is rated to run on different settings which measure different ranges.

Reviewer 3: What is meant by ranges? Frequency ranges?

> The output rate is programmable for both gyroscope and accelerometer.

Reviewer 3: Whats output rate contra sample rate

Author: ```suggestion
The sample rate is programmable for both gyroscope and accelerometer.
```

 \item $\pm 500 ~^{\circ} / s$
 \item $\pm 1000 ~^{\circ} / s$
 \item $\pm 2000 ~^{\circ} / s$
\end{itemize}
\textit{Note: Selecting a higher range will give less accuracy \cite{Mpu-6050-specifications}.}

The accelerometer is rated for measuring gravitational force in different ranges, and it supports the following \cite{Mpu-6050-specifications}:
\begin{itemize}
 \item $\pm 2 ~g$
 \item $\pm 4 ~g$
 \item $\pm 8 ~g$
 \item $\pm 16 ~g$
\end{itemize}
\textit{Note: Selecting a higher range will give less accuracy \cite{Mpu-6050-specifications}.}

The MPU-6050 is able to run in two different output modes to provide data over I$^2$C \cite{Mpu-6050-specifications}. 
\begin{itemize}
 \item Standard mode - 100 KHz - 100 KBits/s
 \item Fast mode - 400 KHz - 400 KBits/s
\end{itemize}

\subsubsection{ESP32}
The ESP32 used is the ESP-WROOM-32 based on ESP32-D0WDQ6 is a dual-core microcontroller based on a Tensilica Xtensa LX6 microprocessor running at 150 MHz to 240 MHz.
The ESP32 supports I$^2$C running in two different modes:
\begin{itemize}
 \item Standard mode - 100 kHz - 100 KBits/s
 \item Fast mode - 400 kHz - 400 KBits/s
\end{itemize}

\subsection{Actual Maximum}

> In \autoref{subsec:theoretical-maximum} the theoretical maximum of the ESP32 and MPU-6050 is researched, to determine where if there is a different actual limit to the amount of samples able to be collected. 

Reviewer 3: This sentence is a little convoluted

Reviewer 3: ```suggestion
In this section the actual maximum of the ESP32 and MPU-6050 is researched, and compared to that of the previous stated theoretical limit.
```

Since Fast mode can not handle 4 times the sample rate of Standard mode, there must be a bottleneck other than the limitations of the I$^2$C protocol.
In this case, since the MPU-6050 is rated to output at 8 kHz and the highest sample rate tested is just 2 kHz, the bottleneck is not with the MPU-6050.
In \autoref{fig:relative-error-percentage} the relative error percentage can be seen for the different sample rates.
It is visible that when the limitation of the I$^2$C protocols Standard mode is reached, the error percentage quickly increases. 
But it is not the same steep slope when the Fast mode error percentage starts to increase.
This could indicate that it is not the same bottleneck for Fast mode and Standard mode.
The test does not directly indicate where the problem is, but it could be speculated that since the ESP32 has to read the data and store it, the ESP32 is the bottleneck.
A possible solution to this is to use a more powerful microcontroller, but this would possibly increase the cost of the product.

> To conclude, the test demonstrated that the max sample rate supported by the system is around 1500 Hz, which will be sufficient for this project.

Reviewer 3: We should maybe state why it is sufficient. If it is.
